Book/php7/zend_engine/zend_opcache.rst

```diff
@@ -1,2 +1,104 @@
+So to store the compiled opcodes for future requests opcache puts them in a shared memory segment (SHM) that can not
+only be accessed by the same process after the request has finished but also by other processes handling other requests.
+This means that if you are using php-fpm with many children they will only store the opcodes for each PHP file once in
+SHM. 
+
+This comes with a significant restriction: The opcodes and other data structures in SHM must not be mutated by any
+child request (at least not unless these changes should also affect the other processes). If any changes to SHM are
+made there is a locking mechanism to avoid data races.
+
+The classes and OParrays (arrays of opcodes, i.e. functions) in SHM are mostly immutable. There are currently the
+following exceptions:
+
+* ``zend_persistent_script.dynamic_members`` stores information about the state of the cached script
+* ``zend_class_entry.inheritance_cache`` is extended if a new combination of parent class and interfaces is encountered
+* ``zend_op.handler`` can be replaced at runtime by the tracing JIT to start executing JITted code
+* ``zend_string.gc.refcount`` for persistent strings can be replaced with map ptr offset to cache class lookups by name
+
+map_ptr
+-------
+
+As mentioned in the introduction, SHM is shared for all child processes. Thus, no changes can be made to the data
+structures living in SHM unless 1. the changes are supposed to be reflected in all child processes and 2. the SHM
+segment has been locked before any changes are written to avoid data races.
+
+Sometimes the data structures contain a field that should be different per process or request. The ``map_ptr`` mechanism
+can be used to achieve this. In short, instead of storing a pointer to per-process data directly (as that would affect
+all processes but each having different addresses for this process specific data), a unique offset is assigned during
+compilation. During runtime, local memory is allocated to hold enough space for all the entries referenced in these
+offsets. Each offset then corresponds to an item in the local memory segment without needing to know the per-process
+location of the exact element.
+
+The map ptr can also store real pointers. This can be useful when the data structure doesn't live in SHM and thus
+doesn't need to store the value in a separate map. The fact that pointers are aligned by their size is used to
+differentiate between offsets and real pointers. The offsets start at 1, 9, 17, etc (assuming a 64-bit pointer size).
+However, pointers (unless padding was removed) would not get aligned this way. Instead, they would be stored in
+addresses that end with 0x0 or 0x8 given that they are 8-byte aligned. The macro ``ZEND_MAP_PTR_IS_OFFSET`` will be used
+internally to check if the ``0b1`` byte is set. If it is, the value stored is an offset. Otherwise it's a direct
```

Comment:
byte -> bit?

Reply:
Thanks! That makes sense. I guess we're also never using `ZEND_MAP_PTR_DEF` to store an offset into a structure.

```c
struct Foo {
    bool a;
    bool b;
};

struct Bar {
    ZEND_MAP_PTR_DEF(bool*, b);
};

struct Foo *foo = emalloc(sizeof(struct Foo));
bool *b = &foo->b; // b would have the 0b1 bit set

struct Bar *bar = emalloc(sizeof(struct Bar));
ZEND_MAP_PTR_INIT(bar->b, b); // bar->b would be interpreted as an offset
```

I'll adjust the explanation accordingly.